nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeduplicateRecords.java

```diff
@@ -0,0 +1,690 @@
+                    recordValue = Joiner.on(JOIN_CHAR).join(record.getValues());
+                } else {
+                    recordValue = executeDynamicRecordPaths(context, record, flowFile);
+                }
+
+                String recordHash = messageDigest != null
+                        ? Hex.encodeHexString(messageDigest.digest(getBytesUtf8(recordValue)))
+                        : recordValue;
+
+                if (!useInMemoryStrategy && context.getProperty(CACHE_IDENTIFIER).isSet()) {
+                    Map<String, String> additional = new HashMap<>();
+                    additional.put("record.hash.value", recordHash);
+                    String rawPath = context.getProperty(CACHE_IDENTIFIER).evaluateAttributeExpressions(flowFile, additional).getValue();
+                    RecordPath compiled = recordPathCache.getCompiled(rawPath);
+                    RecordPathResult result = compiled.evaluate(record);
+                    FieldValue fieldValue = result.getSelectedFields().findFirst().get();
+                    if (fieldValue.getValue() == null) {
+                        throw new ProcessException(String.format("The path \"%s\" failed to create an ID value at record index %d", rawPath, index));
+                    }
+
+                    recordHash = fieldValue.getValue().toString();
+                }
+
+                if (filter.contains(recordHash)) {
+                    duplicatesWriter.write(record);
+                } else {
+                    nonDuplicatesWriter.write(record);
+                    filter.put(recordHash);
+                }
+
+                index++;
+            }
+
+            // Route Non-Duplicates FlowFile
+            final WriteResult nonDuplicatesWriteResult = nonDuplicatesWriter.finishRecordSet();
+            // Route Duplicates FlowFile
+            final WriteResult duplicatesWriteResult = duplicatesWriter.finishRecordSet();
+
+            reader.close();
+            inputStream.close();
+            nonDuplicatesWriter.close();
+            nonDupeStream.close();
+            duplicatesWriter.close();
+            dupeStream.close();
+
+            final boolean includeZeroRecordFlowFiles = context.getProperty(INCLUDE_ZERO_RECORD_FLOWFILES).asBoolean();
+
+            session.adjustCounter("Records Processed",
+                    nonDuplicatesWriteResult.getRecordCount() + duplicatesWriteResult.getRecordCount(), false);
+
+            sendOrRemove(session, duplicatesFlowFile, REL_DUPLICATE, duplicatesWriter.getMimeType(),
+                    includeZeroRecordFlowFiles, duplicatesWriteResult);
+
+            sendOrRemove(session, nonDuplicatesFlowFile, REL_NON_DUPLICATE, nonDuplicatesWriter.getMimeType(),
+                    includeZeroRecordFlowFiles, nonDuplicatesWriteResult);
+
+            session.transfer(flowFile, REL_ORIGINAL);
+
+        } catch (final Exception e) {
+            logger.error("Failed in detecting duplicate records.", e);
```

Comment:
It might be helpful to include the current record index for troubleshooting:
```suggestion
            logger.error("Duplicate record detection failed: current record [{}]", index, e);
```